Sources/ShapedArray/Operations/pack.swift

```diff
@@ -0,0 +1,40 @@
+//
+//  File.swift
+//  
+//
+//  Created by Jaap Wijnen on 10/08/2022.
+//
```

Comment:
For how much I dislike file headers, we'll need to adhere to correct licensing conventions like those by apple [here](https://github.com/apple/swift-numerics/blob/2624562948fd284a3e64ad9e96706f51d9a6a6e9/Sources/Numerics/Numerics.swift#L1-L10) and S4TF [here](https://github.com/s4tf/s4tf/blob/a31f79a7a8e92ff708b1034286a23510b23af5f3/Sources/TensorFlow/Core/ShapedArray.swift#L1-L13) for anything we borrow.

According to Wikipedia 
>it does not require a [derivative work](https://en.wikipedia.org/wiki/Derivative_work) of the software, or modifications to the original, to be distributed using the same license. It still requires application of the same license to all unmodified parts. In every licensed file, original copyright, patent, trademark, and attribution notices must be preserved (excluding notices that do not pertain to any part of the derivative works). In every licensed file changed, a notification must be added stating that changes have been made to that file.

Additionally from Apache 2
> You must retain, in the Source form of any Derivative Works that You distribute, all copyright, patent, trademark, and attribution notices from the Source form of the Work, excluding those notices that do not pertain to any part of the Derivative Works; and

(FYI the next section of the license is about retaining NOTICES, but there are none in S4TF so we can ignore that.)

I interpret this as for all code originally from S4TF we'll need to include the license header and also add to the copyright
```
// Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
such as here https://github.com/swift-numerics-workgroup/swift-shaped-array/blob/3796ba585baef710edd89ecac268e8f47a0d85cd/Sources/ShapedArray/ShapedArray.swift#L1-L13

should _additionally_ include something like
```
// Copyright 2022 SNWG Authors.
// Modified by 2022 SNWG Authors.
```
to specify our copyright and that we modified it. Any files we do not change must just keep the original license.

Does this interpretation seem correct?

Reply:
Also I noticed this was not actually supposed to be modified https://github.com/swift-numerics-workgroup/swift-shaped-array/blob/3796ba585baef710edd89ecac268e8f47a0d85cd/LICENSE#L189-L190
as one can see by example [here in swift-numerics](https://github.com/apple/swift-numerics/blob/2624562948fd284a3e64ad9e96706f51d9a6a6e9/LICENSE.txt#L189)

Reply:
Great catch!

Reply:
> to specify our copyright and that we modified it. Any files we do not change must just keep the original license.
>
> Does this interpretation seem correct?

I think that interpretation is correct. Make a similar header for new files, just replacing TensorFlow with SNWG. Also, let's replace "The TensorFlow Authors" with "The Swift Numerics workgroup". This matches the style of [SSWG](https://www.swift.org/sswg/)'s name. The website workgroup lowercases "website", but I think Numerics should be capitalized. "Workgroup" is lowercased in both instances.

Reply:
For consistency, this is an extension of the Swift for TensorFlow project. They thought up ShapedArray, so they have to intellectual rights. We are just enhancing their code. Using the same header for all files is more consistent, so let's stay with attributing all contributions to TensorFlow. Better safe than sorry.

Reply:
From [Apache 2 4.b](https://www.apache.org/licenses/LICENSE-2.0.html):
>You **must** cause any modified files to carry prominent notices stating that You changed the files; and

Which is why I think we _must also_ include something like
```
// Modified by 2022 SNWG Authors.
```
after the copyright for any file of derivative work (i.e. we split up or modified in some way), and additionally at the end of section 4
>You may add Your own copyright statement to Your modifications

I think implies we _may additionally_ add
```
// Copyright 2022 SNWG Authors.
```
to any file we've edited, but "Better safe than sorry" is a fair statement so we can eschew our own copyright on derivative source. However, anything _new_ we write (i.e. not modifications of theirs) is ours to copyright which implies adding our copyright and _not_ theirs which is inconsistent with your proposal of 
> Using the same header for all files is more consistent

because it seems that we'll actually require 2 different headers depending on the original derived author (us or Tensorflow).

What constitutes _new_ work is new implementation since the following line explicitly calls out that "interfaces" we "link (or bind by name)" i.e. function calls to original or derivative code or references to original or derivative types _do not_ constitute derivative work:
> For the purposes of this License, Derivative Works shall not include works that remain separable from, or merely link (or bind by name) to the interfaces of, the Work and Derivative Works thereof.

Reply:
> What constitutes _new_ work is new implementation since the following line explicitly calls out that "interfaces" we "link (or bind by name)" i.e. function calls to original or derivative code or references to original or derivative types _do not_ constitute derivative work:
> 
> > For the purposes of this License, Derivative Works shall not include works that remain separable from, or merely link (or bind by name) to the interfaces of, the Work and Derivative Works thereof.

That means calling into a pre-compiled system library, as in the linking stage of the Clang compiler. You are not mixing or inlining your code's assembly with the copyright holder's assembly. "Bind by name" is a bit ambiguous, because it could mean:

- (a) Another method to access the copyright holder's source code, which isn't compiler linking. Calling `dlsym` on a system C library is the same mechanism as compiler linking, and is not what "bind by name" refers to. I can't think of a concrete valid example, but you might have some dispatch table mechanism or JIT compilation going on.
- (b) Apple makes a "runtime library exception" in their Apache license so you can embed inlined Swift Stdlib functions in your iOS app binary. "Bind by name" might be similar to the "runtime library exception". This gets murky, as we could just strap `@inline(never)` to our S4TF-defined functions and say it isn't inlined. For some reason, I don't think that will cut it legally.

Our safest bet is to prepend all of our files with the following. Even brand new files take ideas from S4TF, including documentation comments.

```
// Copyright 2019 The TensorFlow Authors. 
// Modified by 2022 The Swift Numerics Workgroup.
```

Reply:
In my fork of S4TF, it's closely tied to TensorFlow, so I don't change the header. When somebody forks a repo on GitHub, they don't peruse every single file and change the header to say "Modified by ...". That would be impractical. Rather, the Git commits document what changes they made. With `swift-shaped-array`, we are starting from scratch with no other way to distinguish:
- (i) Where this source code originated from
- (ii) Which files were modified by whom

Reply:
> "Bind by name" is a bit ambiguous

Likely by intention because it refers to "works" which can be Object or Source, so it's not just Object linking. However US court decided API names are not copyrightable which is why I believe in the case of Source that's almost directly what "interfaces" is referring to, and API's include type names.

"Works" are "Source" and "Object"
>"Source" form shall mean the preferred form for making modifications, including but not limited to software source code, documentation source, and configuration files.

>"Object" form shall mean any form resulting from mechanical transformation or translation of a Source form, including but not limited to compiled object code, generated documentation, and conversions to other media types.

I hadn't consider Object though since we don't distribute any yet, but we might at some point. Thanks for bringing that to light.

In lieu of that I agree
```
// Copyright 2019 The TensorFlow Authors. 
// Modified by 2022 The Swift Numerics Workgroup.
```
everywhere seems most future proof. And I know SO isn't a lawyer but I got a hit that states almost exactly the same thing https://softwareengineering.stackexchange.com/a/220079 .

>"runtime library exception"

Thanks for showing this. After reading https://www.swift.org/community/#runtime-library-exception I'm nearly positive what this is referring to is in an app you will often find hidden deep in settings a "Legal" screen which has every open source library used and their respective licenses stated (Lyft has this as example). This exception states that Swift doesn't need to be included in that section even though the binary Object contains Derivative Work for potentially including Swift runtime. The larger implication is our Object work that mixes with Tensorflow's needs that attribution somewhere too if we ever have Object work.

Reply:
Not sure what the conclusion is here could one of you elaborate?
Also not all code is a literal copy of s4tf I don't think that pack method is from s4tf for example if I remember correctly.

Reply:
> Not sure what the conclusion is here could one of you elaborate?

Copy a header from a file in the S4TF code base. Then, change:

```
// Copyright 2019 The TensorFlow Authors. 
```

To:

```
// Copyright 2019 The TensorFlow Authors. 
// Modified by 2022 The Swift Numerics Workgroup.
```

> Also not all code is a literal copy of s4tf I don't think that pack method is from s4tf for example if I remember correctly.

Yes, but to stay safe, we should apply the header above to all files. Also, some things like the comments and argument labels use ideas from S4TF.

Reply:
Would you mind providing the exact header you'd like to see @philipturner @rexmas ?

Reply:
```swift
// Copyright 2020 The TensorFlow Authors. All Rights Reserved.
// Modified 2022 The Swift Numerics Workgroup.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
```